File: app/src/main/java/n7/mev/main/SoundSourceFactory.java
package n7.mev.main;

import androidx.annotation.NonNull;
import androidx.paging.DataSource;

public class SoundSourceFactory extends DataSource.Factory<Integer, SoundModel> {

    private SoundStorage soundStorage;
    private SoundDataSource soundDataSource;

    public SoundSourceFactory(SoundStorage soundStorage) {
        this.soundStorage = soundStorage;
        soundStorage.setInvalidate(new Invalidate() {
            @Override
            public void invalidate() {
                if (soundDataSource != null) {
                    soundDataSource.invalidate();
                }
            }
        });
    }

    @NonNull
    @Override
    public DataSource<Integer, SoundModel> create() {
        soundDataSource = new SoundDataSource(soundStorage);
        return soundDataSource;
    }

}

File: app/src/main/java/n7/mev/main/SoundSourceFactory.kt
package n7.mev.main

import androidx.paging.DataSource

class SoundSourceFactory(private val soundStorage: SoundStorage) : DataSource.Factory<Int, SoundModel>() {
    private lateinit var soundDataSource: SoundDataSource

    override fun create(): DataSource<Int, SoundModel> {
        soundDataSource = SoundDataSource(soundStorage)
        return soundDataSource
    }

    init {
        soundStorage.setListener(object : Invalidate {
            override fun invalidate() {
                soundDataSource.invalidate()
            }
        })
    }
}
